module/zfs/dmu_tx.c

```diff
@@ -796,11 +798,20 @@ dmu_tx_delay(dmu_tx_t *tx, uint64_t dirty)
 	 */
 	ASSERT3U(dirty, <, zfs_dirty_data_max);
 
-	now = gethrtime();
-	min_tx_time = zfs_delay_scale *
-	    (dirty - delay_min_bytes) / (zfs_dirty_data_max - dirty);
-	min_tx_time = MIN(min_tx_time, zfs_delay_max_ns);
-	if (now > tx->tx_start + min_tx_time)
+	smoothing_time = 0;
+	delay_time = 0;
+
+	if (dirty > delay_min_bytes) {
+		delay_time = zfs_delay_scale *
+		    (dirty - delay_min_bytes) / (zfs_dirty_data_max - dirty);
+	}
+	if (last_smooth > now) {
+		smoothing_time = zfs_smoothing_scale *
+		    (dirty >> 1) / ((delay_min_bytes - dirty) << 1);
```

Comment:
What is the magic with the shifts?  Wouldn't decreasing zfs_smoothing_scale 4 times do the same?

Reply:
If dirty == delay_min_bytes here should be division by zero.  If it goes higher, then smoothing time will be zero due to negative overflow.